In [1]:
import numpy as np
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
df = pd.read_csv('creditcard.csv')
df.head()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0

[5 rows x 31 columns]

In [3]:
df.shape

(284807, 31)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [5]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [8]:
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

#### We can see it is an highly imbalanced dataset with only 0.172% of the transactions being fraudulent.

#### Questions: How to handle imbalanced classes in the dataset? What are some techniques to improve model performance on minority classes?
##### To handle imbalanced classes in the dataset, you can consider the following techniques:
1. **Resampling Techniques**:
   - **Oversampling**: Increase the number of instances in the minority class by duplicating them or generating synthetic samples (e.g., using SMOTE).
   - **Undersampling**: Reduce the number of instances in the majority class to balance the dataset.
2. **Class Weights**: Assign higher weights to the minority class during model training to penalize misclassifications more heavily.
3. **Anomaly Detection**: Treat the minority class as anomalies and use anomaly detection algorithms that are designed to handle imbalanced datasets.
4. **Ensemble Methods**: Use ensemble techniques like Random Forest or Gradient Boosting, which can handle class imbalance better than single models.   

In [9]:
normal = df[df['Class']==0]
fraud = df[df['Class']==1]
print(f"shape of normal transactions: {normal.shape} and shape of fraud transactions: {fraud.shape}")

shape of normal transactions: (284315, 31) and shape of fraud transactions: (492, 31)


In [13]:
normal['Amount'].describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [14]:
fraud['Amount'].describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

**We can clearly see that this data is highly imbalanced containing 284315 records for non-fraud (0) transactions and 492 records for fraud (1) transactions.**

Q. What is Imbalanced Dataset?
- Dataset in which the classes or categories are not represented equally

Q. What problems are caused by highly Imbalanced dataset?
- Biased Model Predictions – The model becomes biased towards the majority class and fails to correctly classify the minority class.
- Poor Recall for Minority Class – The model might achieve high accuracy but perform poorly in detecting the minority class (high false negatives).
- Misleading Performance Metrics – Accuracy can be deceptive; a model predicting only the majority class can still have high accuracy but be useless for real-world applications.

***We will use Under Sampling method to make it balanced. We will take 492 records for fraud (1) transactions and 492 records for non-fraud (0) transactions***

And use this new data frame for model building

In [16]:
df.groupby('Class').mean()

Time        V1        V2  ...       V27       V28      Amount
Class                                    ...                                
0      94838.202258  0.008258 -0.006271  ... -0.000295 -0.000131   88.291022
1      80746.806911 -4.771948  3.623778  ...  0.170575  0.075667  122.211321

[2 rows x 30 columns]

In [18]:
### Resampling Data
from sklearn.utils import resample

In [20]:
normal_under_sample = resample(normal,replace=False,n_samples=len(fraud),random_state=42) # Undersample normal transactions
print("Shape of undersampled data : ",normal_under_sample.shape)

Shape of undersampled data :  (492, 31)


In [21]:
normal_under_sample.head()

Time        V1        V2  ...       V28  Amount  Class
138028   82450.0  1.314539  0.590643  ...  0.031692    0.76      0
63099    50554.0 -0.798672  1.185093  ... -0.112901    4.18      0
73411    55125.0 -0.391128 -0.245540  ...  0.239582   15.00      0
164247  116572.0 -0.060302  1.065093  ...  0.042330   57.00      0
148999   90434.0  1.848433  0.373364  ... -0.042682    0.00      0

[5 rows x 31 columns]

In [24]:
df_new = pd.concat([normal_under_sample,fraud],axis=0) # axis 0 means rows, axis 1 means columns , it means we are concatenating the two dataframes vertically
df_new.shape

(984, 31)

In [25]:
df_new['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [26]:
df_new.to_csv('creditcard_resampled.csv',index=False) # we made index false so that index is not saved in the csv file

In [27]:
df_new.shape

(984, 31)

In [40]:
X = df_new.drop(columns='Class',axis=1)  # All columns except the last one
y = df_new['Class']  # Last column only
print(f"the shape of X is : {X.shape} and the shape of y is : {y.shape}")

the shape of X is : (984, 30) and the shape of y is : (984,)


In [41]:
type(X)

pandas.core.frame.DataFrame

In [42]:
type(y)

pandas.core.series.Series

In [44]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)
print(f"the shape of X_train is : {X_train.shape} and the shape of y_train is : {y_train.shape}")
print(f"the shape of X_test is : {X_test.shape} and the shape of y_test is : {y_test.shape}")

the shape of X_train is : (787, 30) and the shape of y_train is : (787,)
the shape of X_test is : (197, 30) and the shape of y_test is : (197,)


In [ ]:
### Using dagshub and mlflow to track experiments
import mlflow
import dagshub  
mlflow.set_tracking_uri("your Mlflow Tracking URI")  # Replace with your actual MLflow tracking URI
dagshub.init(repo_owner="your repo ownwer", repo_name="Your repo name", mlflow=True)
mlflow.set_experiment("Baseline_Logistic_Regression")


❗❗❗ AUTHORIZATION REQUIRED ❗❗❗

Output()



Open the following link in your browser to authorize the client:
https://dagshub.com/login/oauth/authorize?state=a0b2d4cb-823d-4175-a6c7-9f89d0b9af92&client_id=32b60ba385aa7cecf24046d8195a71c07dd345d9657977863b52e7748e0f0f28&middleman_request_id=70585b78d0bcbf125046f1de8a96cde12ff523ee173b9613a354a2c303fa4267




Accessing as bahuguna.subrat211996

Initialized MLflow to track repo "bahuguna.subrat211996/Fraud-Detection-Project"

Repository bahuguna.subrat211996/Fraud-Detection-Project initialized!

2025/06/28 16:00:09 INFO mlflow.tracking.fluent: Experiment with name 'Baseline_Logistic_Regression' does not exist. Creating a new experiment.


<Experiment: artifact_location='mlflow-artifacts:/26c4b267a1ac44d6bc89dc3e00874d9b', creation_time=1751106609864, experiment_id='0', last_update_time=1751106609864, lifecycle_stage='active', name='Baseline_Logistic_Regression', tags={}>

In [49]:
import mlflow
import logging
import os
import time

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, f1_score, precision_score,confusion_matrix,recall_score

# Set up logging
# here we set up basic configuration for logging , the level is set to INFO, which means that all messages at this level and above will be logged.
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

logging.info("Starting the MLflow experiment")

with mlflow.start_run():
    start_time = time.time()

    try:
        logging.info("loggin preprocessing parameters")
        mlflow.log_param("test_size", 0.2)
        
        logging.info("Training the Logistic Regression model")
        model = LogisticRegression(max_iter=100,solver='lbfgs', penalty='l2', C=1.0, random_state=42) # max_iter is the maximum number of iterations for the solver to converge, solver is the algorithm to use in the optimization problem, penalty is the type of regularization to use, C is the inverse of regularization strength; smaller values specify stronger regularization.

        logging.info("Fitting the model")
        model.fit(X_train, y_train)

        logging.info("logging model parameters")
        mlflow.log_param("C", model.C) # this is the inverse of regularization strength; smaller values specify stronger regularization.
        mlflow.log_param("penalty", model.penalty) # this is the type of regularization to use, such as 'l2' for L2 regularization.
        mlflow.log_param("solver", model.solver) # this is the algorithm to use in the optimization problem, such as 'lbfgs' or 'liblinear'.
        mlflow.log_param("max_iter", model.max_iter) # this is the maximum number of iterations for the solver to converge.

        logging.info("Making predictions on the test set")
        y_pred = model.predict(X_test)

        logging.info("Calculating evaluation metrics")
        accuracy = accuracy_score(y_test, y_pred)
        precision = precision_score(y_test, y_pred) 
        recall = recall_score(y_test, y_pred)
        f1 = f1_score(y_test, y_pred)

        logging.info("Calculating evaluation metrics")
        mlflow.log_metric("accuracy", accuracy)
        mlflow.log_metric("precision", precision)
        mlflow.log_metric("recall", recall)
        mlflow.log_metric("f1_score", f1)

        logging.info("Logging the model")
        # mlflow.sklearn.log_model(model, "model") # log_model(object of model class, artifact_path) -> error 
        mlflow.log_param("model name", "Logistic Regression")

        end_time = time.time()
        elapsed_time = end_time - start_time
        logging.info(f"Experiment completed in {elapsed_time:.2f} seconds")

        logging.info(f"accuracy: {accuracy:.4f}, precision: {precision:.4f}, recall: {recall:.4f}, f1_score: {f1:.4f}")

    except Exception as e:
        logging.error(f"An error occurred: {e}")


2025-06-28 16:42:07,349 - INFO - Starting the MLflow experiment
2025-06-28 16:42:07,725 - INFO - loggin preprocessing parameters
2025-06-28 16:42:08,080 - INFO - Training the Logistic Regression model
2025-06-28 16:42:08,081 - INFO - Fitting the model
c:\Users\Anonymous\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
2025-06-28 16:42:08,137 - INFO - logging model parameters
2025-06-28 16:42:09,502 - INFO - Making predictions on the test set
2025-06-28 16:42:09,502 - INFO - Calculating evaluation metrics
2025-06-28 16:42:09,518 - INFO - Calcula

🏃 View run spiffy-lynx-747 at: https://dagshub.com/bahuguna.subrat211996/Fraud-Detection-Project.mlflow/#/experiments/0/runs/e5d4c99e50d04aa4a1b477e94361132e
🧪 View experiment at: https://dagshub.com/bahuguna.subrat211996/Fraud-Detection-Project.mlflow/#/experiments/0
